cmd/clusterctl/client/repository/components.go

```diff
@@ -439,6 +444,39 @@ func fixCRDWebhookNamespaceReference(o unstructured.Unstructured, targetNamespac
 	return o, errors.Errorf("failed to patch %s CustomResourceDefinition", version)
 }
 
+// fixCertificate fixes the dnsNames of cert-manager Certificates.
+func fixCertificate(o unstructured.Unstructured, targetNamespace string) (unstructured.Unstructured, error) {
+	dnsNames, ok, err := unstructured.NestedStringSlice(o.UnstructuredContent(), "spec", "dnsNames")
+	if err != nil {
+		return o, errors.Wrapf(err, "failed to get .spec.dnsNames from Certificate %s/%s", o.GetNamespace(), o.GetName())
+	}
+	// Return if we don't find .spec.dnsNames.
+	if !ok {
+		return o, nil
+	}
+
+	// Iterate through dnsNames and adjust the namespace.
+	// The dnsNames slice usually looks like this:
+	// - $(SERVICE_NAME).$(SERVICE_NAMESPACE).svc
+	// - $(SERVICE_NAME).$(SERVICE_NAMESPACE).svc.cluster.local
+	for i, dnsName := range dnsNames {
+		dnsNameSplit := strings.Split(dnsName, ".")
+		// Skip this dnsName if the dnsName doesn't have at least 3 segments.
+		if len(dnsNameSplit) < 3 {
+			continue
+		}
+
+		dnsNameSplit[1] = targetNamespace
+		dnsNames[i] = strings.Join(dnsNameSplit, ".")
```

Comment:
could this be a simple replace of ".capi-system." with ".target-namespace."?

Reply:
The question is where do I get the source namespace from. If it's fine to take it from the Certificate itself it would be easy to do.

Reply:
And it's save as long as the namespace name is not part of the service name (which it isn't in our cases)

Reply:
I pushed a new version with replace. Please let me know if you prefer that one (then I'll squash).